File: backend/schemas/schemas.py
# backend/schemas/schemas.py
from pydantic import BaseModel, Field, field_validator, model_validator
from typing import List, Optional
from datetime import date, datetime
from decimal import Decimal

# Huesped Schemas
class HuespedBase(BaseModel):
    nombre: str = Field(..., min_length=3, max_length=255, description="Nombre completo del huésped")
    correo: str = Field(..., pattern=r"^[^@]+@[^@]+\.[^@]+$", description="Correo electrónico único")
    telefono: str = Field(..., pattern=r"^\+?[0-9\s\-]{7,20}$", description="Número de teléfono")
    documento: str = Field(..., min_length=5, max_length=50, description="Documento de identidad único")
    tipo_documento: str = Field(..., pattern="^(DUI|PASAPORTE|CONSTANCIA DE RESIDENCIA)$", description="Tipo de documento")

class HuespedCreate(HuespedBase):
    pass

class HuespedResponse(HuespedBase):
    id_huesped: int

    class Config:
        from_attributes = True

# Detalle Reservacion Schemas
class DetalleReservacionBase(BaseModel):
    id_habitacion: int
    cant_huespedes: int = Field(..., gt=0, description="Huéspedes en esta habitación")
    fecha_entrada: date
    fecha_salida: date

    @model_validator(mode="after")
    def validate_dates(self) -> "DetalleReservacionBase":
        if self.fecha_entrada >= self.fecha_salida:
            raise ValueError("La fecha de entrada debe ser anterior a la fecha de salida.")
        return self

class DetalleReservacionCreate(DetalleReservacionBase):
    pass

class DetalleReservacionResponse(DetalleReservacionBase):
    id_detalle_reservacion: int
    id_reservacion: int

    class Config:
        from_attributes = True

# Reservacion Schemas
class ReservacionCreate(BaseModel):
    id_empleado: int
    id_huesped: int
    cant_huespedes_totales: int = Field(..., gt=0)
    detalles: List[DetalleReservacionCreate]

class ReservacionResponse(BaseModel):
    id_reservacion: int
    id_empleado: int
    id_huesped: int
    cant_huespedes_totales: int
    estado: str

    class Config:
        from_attributes = True

# Consumo Servicio Schemas
class ConsumoServicioCreate(BaseModel):
    id_servicio: int
    id_habitacion: int

class ConsumoServicioResponse(BaseModel):
    id_consumo_servicio: int
    id_servicio: int
    id_habitacion: int
    id_estadia: int
    hora_consumo: datetime

    class Config:
        from_attributes = True

# Estadia Schemas
class EstadiaCreate(BaseModel):
    id_reservacion: int

class EstadiaCheckout(BaseModel):
    id_empleado: int
    metodo_pago: str = Field("EFECTIVO", pattern="^(EFECTIVO|TRANSFERENCIA|TARJETA|BITCOIN|PAYPAL)$")

class EstadiaResponse(BaseModel):
    id_estadia: int
    id_reservacion: int
    checkin: datetime
    checkout: Optional[datetime] = None

    class Config:
        from_attributes = True

# Resenia Schemas
class ReseniaCreate(BaseModel):
    id_huesped: int
    calificacion: Decimal = Field(..., ge=1.0, le=5.0)
    comentario: Optional[str] = None

class ReseniaResponse(BaseModel):
    id_resenia: int
    id_estadia: int
    id_huesped: int
    calificacion: Decimal
    comentario: Optional[str]

    class Config:
        from_attributes = True

# Vista Schemas (para reportes y agregaciones)
class TasaOcupacionResponse(BaseModel):
    mes: str
    tipo_habitacion: str
    total_dias_ocupados: int
    total_dias_disponibles: int
    porcentaje_ocupacion: str

class HuespedesPorHotelResponse(BaseModel):
    id_huesped: int
    huesped: str
    correo: str
    documento: str
    tipo_documento: str
    nombre_hotel: str
    veces_hospedado: int

class DatosGeneralesHotelResponse(BaseModel):
    id_hotel: int
    nombre_hotel: str
    calificacion: Optional[Decimal] = None
    direccion: Optional[str] = None
    niveles_edificios: int
    descripcion: Optional[str] = None
    habitaciones_totales: int
    habitaciones_disponibles: int
    habitaciones_ocupadas: int
    habitaciones_mantenimiento: int
    ganancia_promedio_anual: Decimal
    ganancia_promedio_mensual: Decimal

class NyMResponse(BaseModel):
    pass # No usado

class IngresosMesResponse(BaseModel):
    anio: int
    num_mes: int
    mes: str
    ingresos: Decimal

class InfoGeneralHotelResponse(BaseModel):
    hotel: str
    calificacion: Optional[Decimal] = None
    cant_habitaciones: int
    ganancias: Optional[Decimal] = None

class HabitacionDisponibleResponse(BaseModel):
    id_habitacion: int
    hotel: str
    direccion: Optional[str] = None
    nivel: int
    numero_habitacion: int
    tipo_habitacion: str
    precio: Decimal
    estado: str
    capacidad_maxima: int

class GastoHistoricoHuespedResponse(BaseModel):
    id_huesped: int
    huesped: str
    correo: str
    documento: str
    tipo_documento: str
    gasto_total: Optional[Decimal] = None
    cantidad_facturas: int

class ServiciosMasConsumidosResponse(BaseModel):
    id_tipo_habitacion: int
    tipo_habitacion: str
    id_servicio: int
    tipo_servicio: str
    total_consumos: int

class DiasRestantesReservacionResponse(BaseModel):
    id_reservacion: int
    nombre_huesped: str
    documento_huesped: str
    telefono_huesped: str
    nombre_empleado: str
    cant_huespedes_totales: int
    estado_reservacion: str
    fecha_entrada_proxima: date
    fecha_salida_proxima: date
    dias_para_iniciar: int

class DetalleHabitacionComodidadesResponse(BaseModel):
    id_habitacion: int
    numero_habitacion: int
    precio: Decimal
    estado: str
    capacidad_maxima: int
    tipo_habitacion: str
    total_habitaciones_este_tipo: int
    comodidades: Optional[str] = None
    total_comodidades: int

class TotalHabitacionesTipoResponse(BaseModel):
    tipo_habitacion: str
    total_habitaciones: int

class ConsumoEstadiaResponse(BaseModel):
    id_estadia: int
    nombre: str
    documento: str
    reporte_json: dict

class EmpleadoResponse(BaseModel):
    empleado: str
    documento_de_identidad: str
    telefono: Optional[str] = None
    email: str
    rol_de_trabajo: str
    salario: Decimal

class FacturaCompletaResponse(BaseModel):
    id_factura: int
    fecha: datetime
    metodo_pago: str
    total_a_pagar: Decimal
    id_estadia: int
    nombre_empleado: str
    nombre_huesped: str
    correo_huesped: str
    id_huesped: int
    detalle_factura: list

class ServicioResponse(BaseModel):
    id_servicio: int
    tipo_servicio: str
    precio: Decimal

    class Config:
        from_attributes = True

class EstadiaActivaResponse(BaseModel):
    id_estadia: int
    id_reservacion: int
    checkin: datetime
    nombre_huesped: str
    numero_habitacion: int
    tipo_habitacion: str
    id_habitacion: int

    class Config:
        from_attributes = True

class EmpleadoCreate(BaseModel):
    id_tipo_empleado: int
    nombre: str
    correo: str
    telefono: Optional[str] = None
    dui: str
    salario: Decimal

class FacturaSimplificadaResponse(BaseModel):
    id_factura: int
    nombre_huesped: str
    nombre_empleado: str
    fecha: datetime
    metodo_pago: str
    total_a_pagar: Decimal

    class Config:
        from_attributes = True

class AumentoCostoResponse(BaseModel):
    id_aumento_costo: int
    porcentaje_aumento: Decimal
    fecha_inicio: date
    fecha_fin: date
    nombre_temporada: Optional[str] = None
    activado: bool

    class Config:
        from_attributes = True

class DescuentoResponse(BaseModel):
    id_descuento: int
    porcentaje_descuento: Decimal
    cant_dia_hospedado: Optional[int] = None

    class Config:
        from_attributes = True
# --- CRUD de Hoteles ---
class HotelBase(BaseModel):
    nombre: str = Field(..., max_length=255)
    direccion: Optional[str] = None
    niveles_edificios: int
    calificacion: Optional[Decimal] = None
    descripcion: Optional[str] = None

class HotelCreate(HotelBase):
    pass

class HotelUpdate(BaseModel):
    nombre: Optional[str] = Field(None, max_length=255)
    direccion: Optional[str] = None
    niveles_edificios: Optional[int] = None
    calificacion: Optional[Decimal] = None
    descripcion: Optional[str] = None

class HotelResponseBase(HotelBase):
    id_hotel: int
    class Config:
        from_attributes = True

# --- CRUD de Huéspedes ---
class HuespedUpdate(BaseModel):
    nombre: Optional[str] = Field(None, min_length=3, max_length=255)
    correo: Optional[str] = Field(None, pattern=r"^[^@]+@[^@]+\.[^@]+$")
    telefono: Optional[str] = Field(None, pattern=r"^\+?[0-9\s\-]{7,20}$")
    documento: Optional[str] = Field(None, min_length=5, max_length=50)
    tipo_documento: Optional[str] = Field(None, pattern="^(DUI|PASAPORTE|CONSTANCIA DE RESIDENCIA)$")

# --- CRUD de Empleados ---
class EmpleadoUpdate(BaseModel):
    id_tipo_empleado: Optional[int] = None
    nombre: Optional[str] = None
    correo: Optional[str] = None
    telefono: Optional[str] = None
    dui: Optional[str] = None
    salario: Optional[Decimal] = None

class EmpleadoResponseBase(BaseModel):
    id_empleado: int
    id_tipo_empleado: int
    nombre: str
    correo: str
    telefono: Optional[str] = None
    dui: str
    salario: Decimal
    class Config:
        from_attributes = True

# --- CRUD de Servicios ---
class ServicioCreate(BaseModel):
    tipo_servicio: str = Field(..., max_length=100)
    precio: Decimal

class ServicioUpdate(BaseModel):
    tipo_servicio: Optional[str] = Field(None, max_length=100)
    precio: Optional[Decimal] = None

# --- CRUD de Reservaciones ---
class ReservacionUpdate(BaseModel):
    estado: str = Field(..., pattern="^(PENDIENTE|CONFIRMADA|CANCELADA|FINALIZADA)$")

File: frontend/src/lib/schemas.ts
import { z } from 'zod';

const NumericCoerce = z.union([z.number(), z.string().transform(Number)]);

export const HuespedResponseSchema = z.object({
  id_huesped: z.number(),
  nombre: z.string(),
  correo: z.string(),
  telefono: z.string(),
  documento: z.string(),
  tipo_documento: z.string(),
});

export type Huesped = z.infer<typeof HuespedResponseSchema>;

export const HabitacionDisponibleResponseSchema = z.object({
  id_habitacion: z.number(),
  hotel: z.string(),
  direccion: z.string().optional().nullable(),
  nivel: z.number(),
  numero_habitacion: z.number(),
  tipo_habitacion: z.string(),
  precio: NumericCoerce,
  estado: z.string(),
  capacidad_maxima: z.number(),
});

export type Habitacion = z.infer<typeof HabitacionDisponibleResponseSchema>;

export const ReservacionResponseSchema = z.object({
  id_reservacion: z.number(),
  id_empleado: z.number(),
  id_huesped: z.number(),
  cant_huespedes_totales: z.number(),
  estado: z.string(),
});

export type Reservacion = z.infer<typeof ReservacionResponseSchema>;

export const DiasRestantesReservacionResponseSchema = z.object({
  id_reservacion: z.number(),
  nombre_huesped: z.string(),
  documento_huesped: z.string(),
  telefono_huesped: z.string(),
  nombre_empleado: z.string(),
  cant_huespedes_totales: z.number(),
  estado_reservacion: z.string(),
  fecha_entrada_proxima: z.string(),
  fecha_salida_proxima: z.string(),
  dias_para_iniciar: z.number(),
});

export type DiasRestantesReservacion = z.infer<typeof DiasRestantesReservacionResponseSchema>;

export const EmpleadoResponseSchema = z.object({
  empleado: z.string(),
  documento_de_identidad: z.string(),
  telefono: z.string().optional().nullable(),
  email: z.string(),
  rol_de_trabajo: z.string(),
  salario: NumericCoerce,
});

export type Empleado = z.infer<typeof EmpleadoResponseSchema>;

export const ServicioResponseSchema = z.object({
  id_servicio: z.number(),
  tipo_servicio: z.string(),
  precio: NumericCoerce,
});

export type Servicio = z.infer<typeof ServicioResponseSchema>;

export const EstadiaActivaResponseSchema = z.object({
  id_estadia: z.number(),
  id_reservacion: z.number(),
  checkin: z.string(),
  nombre_huesped: z.string(),
  numero_habitacion: z.number(),
  tipo_habitacion: z.string(),
  id_habitacion: z.number(),
});

export type EstadiaActiva = z.infer<typeof EstadiaActivaResponseSchema>;

export const FacturaSimplificadaResponseSchema = z.object({
  id_factura: z.number(),
  nombre_huesped: z.string(),
  nombre_empleado: z.string(),
  fecha: z.string(),
  metodo_pago: z.string(),
  total_a_pagar: NumericCoerce,
});

export type FacturaSimplificada = z.infer<typeof FacturaSimplificadaResponseSchema>;

export const DetalleFacturaSchema = z.object({
  id_detalle_factura: z.number(),
  concepto: z.string(),
  precio_unitario: NumericCoerce,
  cantidad: z.number(),
  subtotal: NumericCoerce,
  monto_descuento: NumericCoerce,
  monto_aumento: NumericCoerce,
  precio_total: NumericCoerce,
});

export const FacturaCompletaResponseSchema = z.object({
  id_factura: z.number(),
  fecha: z.string(),
  metodo_pago: z.string(),
  total_a_pagar: NumericCoerce,
  id_estadia: z.number(),
  nombre_empleado: z.string(),
  nombre_huesped: z.string(),
  correo_huesped: z.string(),
  id_huesped: z.number(),
  detalle_factura: z.array(DetalleFacturaSchema),
});

export type FacturaCompleta = z.infer<typeof FacturaCompletaResponseSchema>;

export const AumentoCostoResponseSchema = z.object({
  id_aumento_costo: z.number(),
  porcentaje_aumento: NumericCoerce,
  fecha_inicio: z.string(),
  fecha_fin: z.string(),
  nombre_temporada: z.string().optional().nullable(),
  activado: z.boolean(),
});

export type AumentoCosto = z.infer<typeof AumentoCostoResponseSchema>;

export const DescuentoResponseSchema = z.object({
  id_descuento: z.number(),
  porcentaje_descuento: NumericCoerce,
  cant_dia_hospedado: z.number().optional().nullable(),
});

export type Descuento = z.infer<typeof DescuentoResponseSchema>;

export const DatosGeneralesHotelResponseSchema = z.object({
  id_hotel: z.number(),
  nombre_hotel: z.string(),
  calificacion: z.union([z.number(), z.string().transform(Number)]).optional().nullable(),
  direccion: z.string().optional().nullable(),
  niveles_edificios: z.number(),
  descripcion: z.string().optional().nullable(),
  habitaciones_totales: z.number(),
  habitaciones_disponibles: z.number(),
  habitaciones_ocupadas: z.number(),
  habitaciones_mantenimiento: z.number(),
  ganancia_promedio_anual: NumericCoerce,
  ganancia_promedio_mensual: NumericCoerce,
});

export type DatosGeneralesHotel = z.infer<typeof DatosGeneralesHotelResponseSchema>;

export const ServiciosMasConsumidosResponseSchema = z.object({
  id_tipo_habitacion: z.number(),
  tipo_habitacion: z.string(),
  id_servicio: z.number(),
  tipo_servicio: z.string(),
  total_consumos: z.number(),
});
export type ServiciosMasConsumidos = z.infer<typeof ServiciosMasConsumidosResponseSchema>;
